src/datasets/features.py

```diff
@@ -165,7 +167,7 @@ def _cast_to_python_objects(obj: Any) -> Tuple[Any, bool]:
         import torch
 
     if isinstance(obj, np.ndarray):
-        return obj.tolist(), True
+        return numpy_to_pyarrow_listarray(obj).tolist(), True
```

Comment:
Thanks to your changes in the arrow writer, it is now able to take numpy arrays as inputs.
Therefore we don't need to cast the numpy arrays to list here, and we can leave the numpy array as is:

```suggestion
        return obj, False
```

Also feel free to rename this function `cast_to_python_or_numpy`.

Moreover for the pytorch and tensorflow tensors we can do the same: we can just get the numpy array instead of the python lists. It means that we can remove the `.tolist()` at line 172 and 174

Reply:
We can't compare 2 dictionaries if they contain numpy array using `assertDictEqual` (essentially, `dict1==dict2` doesn't work). In place of `self.assertDictEqual(casted_obj, expected_obj)` can we use `np.testing.assert_equal(casted_obj, expected_obj)`? We'll have to do this for all cases where we are casting to numpy array. Since `assertDictEqual` is part of unittest case module I'm a little unsure how to proceed ahead

Reply:
Sure you can use `np.testing.assert_equal`. Also please note that `np.testing.assert_equal` won't check for the dtypes of the numpy arrays, which can be an issue.

Reply:
Then we can use `np.testing.assert_equal` to check if they're same and individually check dtype of `casted_obj["col_1"][0]["vec"]` and `casted_obj["col_2"]` which should match that of `expected_obj["col_1"][0]["vec"]` and `expected_obj["col_2"]`?